Add filtering support, rename endpoints, refactor object extend (closes #17)

cdhorn opened the pull request:
@DavidMStraub @Nick-Hall I think this may be ready to merge if it passes muster. It does the following:

1. Renames all the endpoints to their plural form as they represent collections and not singular objects. Note I did not rename the source files in the resources directory, those are still singlar. ie person.py.
2. Adds support for the filtering framework by adding the /filters/{{namespace}} endpoint with the ability to get/create/update/delete custom filters and adding the filter= and rules= query parameters to the people, families, events, places, citations, sources, repositories, media and notes endpoints. filter= take the name of a custom filter. rules= takes a JSON structured argument to build a filter dynamically. 
3. Refactors object_extend to eliminate replicated code.
4. Refactors some of the boolean query parms into strings so the presence of the keyword is sufficient. ie profile, extend, strip.
5. Updates apispec.yaml to reflect the above changes and fix a couple issues.

This again tweaks the test cases to pass but does not add new ones. Once this is in and the example.gramps is merged I plan on focusing on test cases for everything done to date.

DavidMStraub commented:
Thanks!

Again, please try to separate features/changes in the future. E.g. having the renaming of *all* endpoints plus the massive filtering feature in a single PR makes it hard to review.

cdhorn commented:
Yes, I know, going forward I will try to keep them smaller and more specific like you did with the example.gramps work.

DavidMStraub commented on gramps_webapi/api/resources/base.py:
```diff
@@ -65,22 +69,24 @@ def get(self, args: Dict, handle: str) -> Response:
         try:
```
This method is getting really long; it would be more readable by putting the various `if` cases into separate methods. But I leave it up to you.

cdhorn replied:
This will need to be refactored for sorting so can address it then.  I think I agree with you, handle goes away. With the HasIdOf rule the gramps_id is not needed either.

DavidMStraub commented on gramps_webapi/api/resources/emit.py:
```diff
@@ -26,19 +26,22 @@ def __init__(self):
-        # pylint:disable=dangerous-default-value
+    def response(
+        self, status: int = 200, payload: Any = {}, args: Dict = {}
+    ) -> Response:
```
*Never* use an empty dict as a default function argument. It is mutable and any modification that you make will persist. (See e.g. [here](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python)).

Use:

```python
def f(args: Optional[Dict] = None):
    args = args or {}
```

cdhorn replied:
I will fix this.

DavidMStraub commented on gramps_webapi/api/resources/event.py:
```diff
@@ -24,9 +24,9 @@ class EventResourceHelper(GrampsObjectResourceHelper):
         """Extend event attributes as needed."""
         db_handle = self.db_handle
-        if args["profile"]:
+        if "profile" in args:
```
Why is this no longer a boolean? Will `profile=0` be interpreted as `True`?

cdhorn replied:
If profile is not present in the parms then it is not generated, if ?profile is in the response it is generated. No need for =0 or =1. I think it is cleaner. Same for the others that were booleans.

DavidMStraub commented:
I looked at everything *except* the filtering feature and it looks good! Only a couple of minor comments.

Concerning the filters:
- Can you give one or two example URLs to make it easier to understand how it works?
- Do you plan to add unit tests?

cdhorn commented:
Well, that's embarassing, indented something too far. Please use latest, I also rebased and fixed the empty dict default arg. 

I included what I hoped was a clear and concise description for the rules= query parm in the apispec.yaml in this PR, please review it and let me know if you think it is sufficient or could be worded better.

Before starting examine the filters people endpoint and look over the available rules:
`api/filters/people`

This constructs a filter for a specific gramps_id, hence the reason I think we no longer need it although it is a convenient "shorthand":
`api/people/?rules={"r":[{"n":"HasIdOf","v":["P1"]}]}`

This constructs a filter on all males with the tag named "DNA Match" who are marked private:
`api/people/?rules=[{"n":"HasTag","v":["DNA Match"]},{"n":"IsMale"},{"n":"PeoplePrivate"}]}`

This constructs a filter to return people matching tag "Direct Ancestor" or "Military":
`api/people/?rules={"f":"or","r":[{"n":"HasTag","v":["Direct Ancestor"]},{"n":"HasTag","v":["Military"]}]}`

This constructs a filter to return all female ancestors of person with gramps_id P1:
`api/people/?rules={"r":[{"n":"IsAncestorOf","v":["P1"]},{"n":"IsFemale"}]}`

Let us suppose we used POST and created a named custom filter "FemaleAncestors" with the above filter. (Note for POST it expects full keywords in the JSON and not the shorthand characters used in the query parm to help shorten it)

This then would let us run the custom filter directly as Nick suggested:
`api/people/?filter=FemaleAncestors`

This then uses that custom filter in another filter we are constructing to fetch all descendents of the female ancestors of the person:
`api/people/?rules={"r":[{"n":"IsDescendantOfFilterMatch","v":["FemaleAncestors"]}]}`

The thought occured to me we might provide a set of "shorthand" query parms for common filters, like the tags= example, but with such rich functionality available with rules= are they really needed?


Regarding unit tests, as I previously mentioned my plan is to back track and focus on unit tests for **_all_** of the existing endpoints **_and_** query parms, including everything in this PR, as I said I would do so and see that as my responsibility.  Rather than one big PR I will try to submit those in smaller chunks, I think an endpoint at a time for the major objects and then one final one for all the loose ends makes the most sense.

There will probably be one small PR before then once this is merged to modify the extend query parm to accept a optional list of specific keys to process. For example, extend with no arguments currently defaults to everything, but extend=events,media on a person would only return those two. I did not think of doing that when I refactored it and it is so obvious it should be done that way in hindsight to make it more flexible.

The reason I would submit that one before diving into test cases is so as I work through the unit tests for each end point I can include the necessary cases for that as well. I considered adding it to this PR as I did refactor that code as part of this but decided it could just be a quick follow on. 

Let me know if my current plans sound agreeable to you. Actually I will open the test case issues tonight.

DavidMStraub commented:
> I included what I hoped was a clear and concise description for the rules= query parm in the apispec.yaml in this PR, please review it and let me know if you think it is sufficient or could be worded better.

Sorry for having missed that!

> Before starting examine the filters people endpoint and look over the available rules:
> `api/filters/people`
> 
> This constructs a filter for a specific gramps_id, hence the reason I think we no longer need it although it is a convenient "shorthand":
> `api/people/?rules={"r":[{"n":"HasIdOf","v":["P1"]}]}`
> 
> This constructs a filter on all males with the tag named "DNA Match" who are marked private:
> `api/people/?rules=[{"n":"HasTag","v":["DNA Match"]},{"n":"IsMale"},{"n":"PeoplePrivate"}]}`
> 
> This constructs a filter to return people matching tag "Direct Ancestor" or "Military":
> `api/people/?rules={"f":"or","r":[{"n":"HasTag","v":["Direct Ancestor"]},{"n":"HasTag","v":["Military"]}]}`
> 
> This constructs a filter to return all female ancestors of person with gramps_id P1:
> `api/people/?rules={"r":[{"n":"IsAncestorOf","v":["P1"]},{"n":"IsFemale"}]}`

That's really cool, I like it!

What I don't quite understand is the rationale for the shorthands ... given the response will typically be orders of magnitude larger than the request, bandwidth cannot be an issue and I don't think it is more readable either. I also don't think there is any relevant limit on the length of the query string. And the code would be much simpler without translating between short and long forms. Why did you introduce it?

> Let us suppose we used POST and created a named custom filter "FemaleAncestors" with the above filter. (Note for POST it expects full keywords in the JSON and not the shorthand characters used in the query parm to help shorten it)
> 
> This then would let us run the custom filter directly as Nick suggested:
> `api/people/?filter=FemaleAncestors`
> 
> This then uses that custom filter in another filter we are constructing to fetch all descendents of the female ancestors of the person:
> `api/people/?rules={"r":[{"n":"IsDescendantOfFilterMatch","v":["FemaleAncestors"]}]}`

Yes, this is great. I think those are two separate and both very well motivated use cases:

1. An app user creating a custom filter, like in desktop Gramps, and using it in reports/searches/other filters
2. The app itself using the filter to request a subset of items, e.g. for filtering a long list or for displaying search results.

In the second case, the `GET` endpoint is perfect since it doesn't create anything persistent, while the `POST` endpoint implements the first case.

> The thought occured to me we might provide a set of "shorthand" query parms for common filters, like the tags= example, but with such rich functionality available with rules= are they really needed?

Yes, good point. Let's leave it out for the moment in order not to complicate things. Probably it's not necessary anyway.

> Regarding unit tests, as I previously mentioned my plan is to back track and focus on unit tests for **_all_** of the existing endpoints **_and_** query parms, including everything in this PR, as I said I would do so and see that as my responsibility. Rather than one big PR I will try to submit those in smaller chunks, I think an endpoint at a time for the major objects and then one final one for all the loose ends makes the most sense.

:+1:

DavidMStraub commented on gramps_webapi/api/resources/filters.py:
```diff
@@ -0,0 +1,298 @@
+from flask import Response, abort
+from gramps.gen.db.base import DbReadBase
+from gramps.gen.filters import GenericFilter
+from marshmallow import Schema, ValidationError, fields
```
Have you forgotten adding `marshmallow` to `setup.py`? I don't know if there are any tests importing this module yet...

cdhorn replied:
It is required for webargs, do you think it needs to be added separately?

DavidMStraub commented on gramps_webapi/api/resources/filters.py:
```diff
@@ -0,0 +1,298 @@
+        location="query",
+    )
+    def get(self, args: Dict, namespace: str) -> Response:
+        """Get available custom filters and rules."""
```
This is a loooong method. Possible to refactor? We can also do it later, but also for unit tests it's difficult having so long methods where we cannot test individual pieces.

cdhorn replied:
I'm not sure criteria for what is considered long. I guess the two functions it performs could be broken out and called separately.

DavidMStraub commented on gramps_webapi/api/resources/filters.py:
```diff
@@ -0,0 +1,298 @@
+        except KeyError:
+            abort(404)
+
+        new_filter = build_filter(args, namespace, keys="long")
```
A function argument that only has two valid values (in this case "long" and "short") cries out being a boolean rather than a string where all except two values are undefined. `long_keys=True`.

cdhorn replied:
Yes, good point. That will go away with a refactor.

DavidMStraub commented on gramps_webapi/api/resources/filters.py:
```diff
@@ -0,0 +1,298 @@
+        },
+        location="json",
+    )
+    def delete(self, args: Dict, namespace: str) -> Response:
```
It's great that you're using type hints everywhere, it makes the code much more easy to understand.

For dicts, it's also helpful to indicate the type of the keys (and possibly values); this also catches possible bugs early in my experience. In this case, problably `Dict[str, str]`.

cdhorn replied:
Yes, I am trying to get in the practice of doing so, they are helpful. Good idea on the Dict hints, did not occur to me, I will try to do that as well.

DavidMStraub commented on gramps_webapi/api/resources/filters.py:
```diff
@@ -0,0 +1,298 @@
+        return self.response(201, {"message": "Added filter: " + new_filter.get_name()})
+
+    @use_args(FilterSchema(), location="json")
+    def put(self, args: Dict, namespace: str) -> Response:
```
Another long method.

DavidMStraub commented:
Done reviewing :slightly_smiling_face:

cdhorn commented:
> What I don't quite understand is the rationale for the shorthands ... Why did you introduce it?

In all honesty I had the query parms worked out first and thought keeping them short made sense, and then decided to use the full keywords when I got around to the other code. I agree it makes sense to get rid of them, it is less code and probably clearer. I will refactor in a little while and push a commit doing so later today.

With rules= functionality do you object to me removing gramps_id= and handle= from GrampsObjectsResource now as well when I do so?

DavidMStraub commented:
> In all honesty I had the query parms worked out first and thought keeping them short made sense, and then decided to use the full keywords when I got around to the other code. I agree it makes sense to get rid of them, it is less code and probably clearer. I will refactor in a little while and push a commit doing so later today.

:+1:

> With rules= functionality do you object to me removing gramps_id= and handle= from GrampsObjectsResource now as well when I do so?

handle definitely, for this we have a dedicated endpoint.

For `gramps_id`, I am more worried about performance. I find it quite likely that an app might choose to call individual objects by their Gramps IDs rather than handles (e.g. this is how my existing app does it). And while the filter rule with `HasIdOf` and an implementation using `get_person_from_gramps_id` (taking person as an example) return the same result, their implementations are veryy different: the filter implementation will fetch **all** people in the database, construct `Person` instances for each of them, iterate through the created instances and return a single one. The `..._from_gramps_id` method will instead use a database query with `WHERE` clause, which I expect to be *much* faster (even though I haven't tried it and even though the default SQLite DB doesn't have any indices, so the full table must be read anyway). But admittedly from an API point of view it doesn't make much sense to duplicate this functionality.

DavidMStraub commented:
And perhaps we shouldn't do premature optimization at the expense of transparency :slightly_smiling_face:

cdhorn commented:
Looked closer at that, yes you are right there is a definite benefit to keeping gramps_id as is. I will pull handle out though, as you previously mentioned it is unnecessary.

Nick-Hall commented:
> The ..._from_gramps_id method will instead use a database query with WHERE clause, which I expect to be much faster (even though I haven't tried it and even though the default SQLite DB doesn't have any indices, so the full table must be read anyway).

The `gramps_id` columns are indexed and these methods will be fast.

cdhorn commented:
Also refactored get method in FilterResource class.

DavidMStraub commented:
So, all set already?

cdhorn commented:
If everything passes muster with you, sure! :-)
